### PR TITLE
flow matching lcm scheduler

### PR DESCRIPTION
**Pull request by quickjkee**
## Why?

Most cutting-edge image generative models (<a href="https://github.com/Stability-AI/sd3.5?ysclid=m8sp05sq5j831299212">SD3.5</a>, 
<a href="https://github.com/black-forest-labs/flux?ysclid=m8sp107j7777915043">FLUX</a>) are formulated using <a href="https://arxiv.org/abs/2403.03206">flow matching</a>. Moreover, some state-of-the-art distillation methods (<a href="https://github.com/tianweiy/DMD2/tree/main">DMD2</a>, <a href="https://github.com/luosiallen/latent-consistency-model">CD</a>) use stochastic multistep sampling (<a href="https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_lcm.py">LCM scheduler</a>). However, there is no realization of the LCM scheduler for the flow matching formulation. 

## What does this PR do?

Thus, this PR provides a ```scheduling_flow_match_lcm.py``` that is in the same spirit as the <a href="https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_lcm.py">scheduling_lcm.py</a>. The goal is for distilled models (using, e.g <a href="https://github.com/Stability-AI/sd3.5?ysclid=m8sp05sq5j831299212">SD3.5</a> or <a href="https://github.com/black-forest-labs/flux?ysclid=m8sp107j7777915043">FLUX</a>) to use this sampler. 

Moreover, as an additional tool, this PR provides the ability to generate on different resolution scales. 
Which could significantly accelerate the generation process. This concept was introduced in <a href="https://arxiv.org/pdf/2503.16397">SwD</a>.

## How it works?

For example, one of the most prominent distillation approaches, <a href="https://github.com/tianweiy/DMD2/tree/main">DMD2</a>, uses <a href="https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_lcm.py">LCM</a>. It is distilled from <a href="https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0">SDXL</a>. But there is also an <a href="https://huggingface.co/yresearch/stable-diffusion-3.5-large-dmd2">implementation</a> for <a href="https://huggingface.co/stabilityai/stable-diffusion-3.5-large">SD3.5 Large</a> and with PR it works as follows:

```python
import torch
from diffusers.schedulers.scheduling_flow_match_lcm import FlowMatchLCMScheduler
from diffusers import StableDiffusion3Pipeline
from peft import PeftModel

# Loading models
pipe = StableDiffusion3Pipeline.from_pretrained("stabilityai/stable-diffusion-3.5-large", 
                                                torch_dtype=torch.float16)
pipe = pipe.to("cuda")
lora_path = 'yresearch/stable-diffusion-3.5-large-dmd2'
pipe.transformer = PeftModel.from_pretrained(
    pipe.transformer,
    lora_path,
)

# LCM scheduler 
pipe.scheduler = FlowMatchLCMScheduler.from_config(pipe.scheduler.config, 
                                                   shift=1.0)

# Generation
image = pipe(
    "a cat reading a newspaper",
    sigmas=[1.0000, 0.9454, 0.8959, 0.7904, 0.7371, 0.6022],
    guidance_scale=0.0,
    generator=torch.Generator().manual_seed(0),
).images[0]
```

To generate in a scale-wise regime, it is necessary to provide scale factors through the ```pipe.scheduler.set_scale_factors```. 
This modification is optional.

```python
import torch
from diffusers.schedulers.scheduling_flow_match_lcm import FlowMatchLCMScheduler
from diffusers import StableDiffusion3Pipeline
from peft import PeftModel

# Loading models
pipe = StableDiffusion3Pipeline.from_pretrained("stabilityai/stable-diffusion-3.5-large", 
                                                torch_dtype=torch.float16)
pipe = pipe.to("cuda")
lora_path = 'yresearch/swd-large-6-steps'
pipe.transformer = PeftModel.from_pretrained(
    pipe.transformer,
    lora_path,
)

# LCM scheduler
pipe.scheduler = FlowMatchLCMScheduler.from_config(pipe.scheduler.config, 
                                                   shift=1.0)

# Setting up the scale factors
# We define it wrt input size 
# In this case: 32->48->64->80->96->128
pipe.scheduler.set_scale_factors(
    scale_factors=[1.5, 2., 2.5, 3., 4.], 
    upscale_type='bicubic'
)

# Generation
image = pipe(
    "a cat reading a newspaper",
    sigmas=[1.0000, 0.9454, 0.8959, 0.7904, 0.7371, 0.6022],
    guidance_scale=0.0,
    generator=torch.Generator().manual_seed(0),
    width=256,
    height=256
).images[0]
```

### CODE REVIEW

**Comment by hlky** [comment]
@bot /style

**Comment by yiyixuxu** [comment]
thanks @quickjkee 
do we have a training script using this scheduler?

**Comment by quickjkee** [comment]
> thanks @quickjkee do we have a training script using this scheduler?

thanks for the question and reviewing @yiyixuxu!
I don't think so. This scheduler is only needed for inference, as the basic LCM scheduler.

**Comment by yiyixuxu** [comment]
@quickjkee 
do you have a training script for these checkpoints that can be used for this scheduler? (even though the training itself does not require it)

**Comment by ShuyUSTC** [comment]
Thank you for your great contribution! Is there any plan to merge this pr? Our [HiDream-I1](https://github.com/HiDream-ai/HiDream-I1) `fast` and `dev` model use this `flow_match_lcm` scheduler by default. We are waiting for this pr to be merged. Thanks!

File: src/diffusers/schedulers/scheduling_flow_match_lcm.py [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+                x0_pred = torch.nn.functional.interpolate(x0_pred, size=size, mode=self._upscale_mode)
+
+        noise = randn_tensor(x0_pred.shape, generator=generator, device=x0_pred.device, dtype=x0_pred.dtype)
+        prev_sample = (1 - sigma_next) * x0_pred + sigma_next * noise
+
+        # upon completion increase step index by one
+        self._step_index += 1
+        if per_token_timesteps is None:
+            # Cast sample back to model compatible dtype
+            prev_sample = prev_sample.to(model_output.dtype)
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_flow_match_lcm.py:480]:
```suggestion
        # Cast sample back to model compatible dtype
        prev_sample = prev_sample.to(model_output.dtype)
```

File: src/diffusers/schedulers/scheduling_flow_match_lcm.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+                A current instance of a sample created by the diffusion process.
+            s_churn (`float`):
+            s_tmin  (`float`):
+            s_tmax  (`float`):
+            s_noise (`float`, defaults to 1.0):
+                Scaling factor for noise added to the sample.
+            generator (`torch.Generator`, *optional*):
+                A random number generator.
+            per_token_timesteps (`torch.Tensor`, *optional*):
+                The timesteps for each token in the sample.
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_flow_match_lcm.py:426]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_flow_match_lcm.py [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+        self,
+        model_output: torch.FloatTensor,
+        timestep: Union[float, torch.FloatTensor],
+        sample: torch.FloatTensor,
+        s_churn: float = 0.0,
+        s_tmin: float = 0.0,
+        s_tmax: float = float("inf"),
+        s_noise: float = 1.0,
+        generator: Optional[torch.Generator] = None,
+        per_token_timesteps: Optional[torch.Tensor] = None,
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_flow_match_lcm.py:404]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_flow_match_lcm.py [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+                The direct output from learned diffusion model.
+            timestep (`float`):
+                The current discrete timestep in the diffusion chain.
+            sample (`torch.FloatTensor`):
+                A current instance of a sample created by the diffusion process.
+            s_churn (`float`):
+            s_tmin  (`float`):
+            s_tmax  (`float`):
+            s_noise (`float`, defaults to 1.0):
+                Scaling factor for noise added to the sample.
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_flow_match_lcm.py:422]:
```suggestion
```

File: src/diffusers/schedulers/scheduling_flow_match_lcm.py [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+
+    def step(
+        self,
+        model_output: torch.FloatTensor,
+        timestep: Union[float, torch.FloatTensor],
+        sample: torch.FloatTensor,
+        s_churn: float = 0.0,
+        s_tmin: float = 0.0,
+        s_tmax: float = float("inf"),
+        s_noise: float = 1.0,
```
Comment (hlky) [inline src/diffusers/schedulers/scheduling_flow_match_lcm.py:402]:
```suggestion
```

**Review by hlky — APPROVED** [review]
Hi @quickjkee. Left a few suggestions on unused parameters but LGTM, thanks!
